pandas/core/indexes/base.py

```diff
@@ -2658,41 +2658,38 @@ def _union(self, other, sort):
         lvals = self._values
         rvals = other._values
 
-        if sort is None and self.is_monotonic and other.is_monotonic:
-            try:
+        try:
+            if sort is None and self.is_monotonic and other.is_monotonic:
                 result = self._outer_indexer(lvals, rvals)[0]
-            except TypeError:
-                # incomparable objects
-                result = list(lvals)
-
-                # worth making this faster? a very unusual case
-                value_set = set(lvals)
-                result.extend([x for x in rvals if x not in value_set])
-                result = Index(result)._values  # do type inference here
-        else:
-            # find indexes of things in "other" that are not in "self"
-            if self.is_unique:
-                indexer = self.get_indexer(other)
-                indexer = (indexer == -1).nonzero()[0]
             else:
-                indexer = algos.unique1d(self.get_indexer_non_unique(other)[1])
-
-            if len(indexer) > 0:
-                other_diff = algos.take_nd(rvals, indexer, allow_fill=False)
-                result = concat_compat((lvals, other_diff))
-
-            else:
-                result = lvals
-
-            if sort is None:
-                try:
-                    result = algos.safe_sort(result)
-                except TypeError as err:
-                    warnings.warn(
-                        f"{err}, sort order is undefined for incomparable objects",
-                        RuntimeWarning,
-                        stacklevel=3,
-                    )
+                # We calculate the sorted union and resort it afterwards
+                result = self._outer_indexer(np.sort(lvals), np.sort(rvals))
+        except TypeError:
+            # incomparable objects
+            result = list(lvals)
+
+            # worth making this faster? a very unusual case
+            value_set = set(lvals)
+            result.extend([x for x in rvals if x not in value_set])
+            result = Index(result)._values  # do type inference here
+        else:
+            if sort is False or not self.is_monotonic or not other.is_monotonic:
+                indexer = []
+                counts = dict(zip(*np.unique(result[0], return_counts=True)))
+                unique_array = algos.unique(np.append(lvals, rvals))
+                # Create indexer to resort result
+                for i, value in enumerate(unique_array):
+                    indexer += [i] * counts[value]
+                result = unique_array.take(indexer)
+        if sort is None and (not self.is_monotonic or not other.is_monotonic):
```

Comment:
is the relevant thing monotonicity or uniqueness?  IIRC Index.is_monotonic is aliased to is_monotonic_increasing

Reply:
If one of both is not monotonic. we have to resort after the original order. Does not matter if they are not unique.